### PR TITLE
FBX loader: full format coverage, animation fidelity, DCC materials, NURBS, constraints

### PR DESCRIPTION
**Pull request by Pryme8**
## The short version

FBX is still the file everybody actually has. Every Maya, Max, MotionBuilder, Blender and Mixamo pipeline spits
one out, and half the models people try to drag into the sandbox are FBX. The pure TS loader that landed in 9.26
was a great start, so I spent some time seeing how far it could go. This PR takes it from "loads a lot of files"
to "loads all of them, and they look and move like they did in the DCC".

Nothing changes for existing users: the defaults are the 9.26 behaviour, and everything new sits behind options,
bundled as `preset: "full"` for people who want all of it in one line.

I did not want to guess at any of this, so every change was checked against ufbx (the C FBX parser that has
feature parity with the Autodesk SDK) on its test corpus: 683 real files from Maya, 3ds Max, MotionBuilder,
Blender and ZBrush, versions FBX 3000 through 7700, ASCII and binary. Before: 501 of those files loaded.
After: all of them, with transforms, triangle counts, animation samples, camera FOVs and lights matching ufbx
to the third decimal.

## What changed

**File formats.** FBX 6.0/6.1 was only partly working (transform animation loaded, but lights, cameras, skins
and blend shapes were silently dropped). FBX 5.x and older did not load at all. Both work now: the old files get
rewritten into the 6.x layout on load so the rest of the loader does not have to care. Big-endian binaries,
every weird ASCII number spelling (`1.#INF`, `nan(ind)`, `&quot;`), files with no `Objects` block, cameras that are
never connected to anything: all handled.

**Animation.** This is where most of the work went. Curves keep their real tangents (Bezier, TCB, auto,
weighted, broken), pre/post extrapolation works, animation layers blend the way MotionBuilder blends them
(additive and override, accumulation modes, animated layer weights), step keys stay stepped, negative clip starts
are fine, and the file's frame rate is used instead of 30. Node transforms are built from the exact FBX TRS chain
with pivots and offsets instead of decomposing matrices, and the RrSs / Rrs inherit modes work for both plain
nodes and rig bones. Visibility, camera, light, material, blend shape weight and even user property animation
come through as Babylon animations. Clips that span days (yes, the corpus has those) used to bake millions of
keys; sampling is now capped per key segment and they load in milliseconds.

**Materials.** Every material flavour the big DCCs write: Lambert/Phong, Maya Standard Surface (Arnold and
OSL), 3ds Max Physical, 3ds Max PBR (metal/rough and spec/gloss), glTF, OpenPBR, Stingray PBS. They all resolve
into one material model and come out as `StandardMaterial` by default, or as `PBRMaterial` for the PBR ones with
`materials: "auto"` (or force either). Texture wrap modes and the old 6.x "texture per layer element" assignment
work too. Layered textures are approximated by their first layer and reported; Babylon has no layer blending to
map them onto.

**Geometry.** Vertex welding (the old loader produced one vertex per polygon corner), optional geometry sharing
between instanced models, an optional unit scale, and the missing geometry types: NURBS surfaces get tessellated
into meshes with the same subdivision the file asks for (a skin or blend shape bound to one is reported and left
off, since the tessellated mesh has no control point mapping), NURBS curves and `Line` geometry become
`LinesMesh` when asked. LOD groups map to Babylon LOD levels honouring each level's show/hide/use-LOD mode,
display layers hide their members, selection sets and user properties land in metadata. The ear clipper was cubic in vertex count and took 82 seconds on a maze made of huge n-gons; it is
now a linked-ring clipper that does the same file in 2 seconds with identical output.

**Cameras and lights.** Field of view now follows the SDK's aspect, aperture and gate fit rules (it matches to
the millidegree across every camera in the corpus), roll and ortho size are right, lights get their range from
far attenuation, and with `attachCamerasAndLights` both follow the node they belong to when it is animated.

**Constraints.** Aim, parent, position, rotation and scale constraints are recorded as metadata, and with
`constraints: "apply"` solved every frame: one solver per scene brackets the scene's animation phase (restores the
unconstrained transforms before animations run, captures and solves after), orders the constraints so a constraint
whose target (or parent) is driven by another one is solved after it, reports cycles, blends from what animation
produced this frame (so a 50% weight stays 50%, whatever value animation wrote), and allocates nothing per frame.
IK stays metadata.

**Diagnostics.** Anything the loader has to skip or approximate is reported through an `onWarning` option and
stored as `fbxDiagnostics` on the root node. Nothing fails silently anymore, and nothing that does work is
reported as unsupported.

## How it was tested

- The ufbx corpus harness (in my repo, linked below): loads all 683 files with a `NullEngine` and compares
  against a JSON dump from ufbx itself. 673 of 673 valid files load, triangle counts match on 672, rest
  transforms match on 3920 of 3929 nodes, animation samples on 4358 of 4370, cameras and lights on all of them.
  The leftovers are deliberate conventions and synthetic torture files (duplicate ids, names made of quotes, NaN
  transforms), documented in the tests.
- Babylon's own 18 FBX visualization playgrounds, replayed with this loader on WebGL2 and pixel-diffed against
  the reference images in the repo: all 18 within 0.5% of pixels (antialiasing noise). This is the check that
  caught the one real bug the headless harness could not: parenting cameras to their nodes put them under the
  root's handedness flip and rendered a mirrored image. Fixed with a pivot, and there is a unit test for it now.
- Every corpus file rendered on a real GPU: no exceptions, no shader errors.
- `npm run build:source`, `npm run lint:check`, the TypeDoc check and the side-effects manifest are clean. The 109
  FBX unit tests in this repo pass: 80 existing (two with expectations updated, see below) and 29 new ones that
  pin every fix and every public option in this PR (ids beyond 2^53, negative clips, Properties60 layouts, legacy
  GlobalSettings, the constraint solver, layered property animation, LOD display modes, NURBS deformers, texture
  creation, diagnostics, presets).
- 478 vitest tests in my repo, including a copy of this repo's FBX unit tests, parser edge cases, feature tests
  and 43 corpus fixtures against oracle dumps.

## Things reviewers should know

Defaults are the 9.26 behaviour. Everything that changes what a scene looks like is an option, and
`preset: "full"` sets them all at once:

| Option                   | default (`"compatible"`) | `preset: "full"` |
| ------------------------ | ------------------------ | ---------------- |
| `materials`              | `"standard"`             | `"auto"`         |
| `shareGeometry`          | `false`                  | `true`           |
| `curves`                 | `"skip"`                 | `"lines"`        |
| `constraints`            | `"metadata"`             | `"apply"`        |
| `rebaseAnimations`       | `true`                   | `false`          |
| `attachCamerasAndLights` | `false`                  | `true`           |

Other things worth knowing:

- New metadata keys on nodes, meshes, materials, cameras and lights, all prefixed `fbx`.
- Object ids beyond 2^53 are accepted (real exporters write them) and kept distinct: the parsers keep the exact
  digits and the connection resolver keys them by that text, so two ids that round to the same double never
  collide. Synthetic ids for legacy objects are allocated outside the ids the file uses.
- Two unit test expectations changed: the unsafe-id tests above, and a layer with override blending at a 50%
  weight is evaluated instead of flagged as unsupported.
- Joint children stay in the node hierarchy instead of being attached to bones.

Not in this PR: IK solving, stereo cameras, audio clips, cache deformers, NURBS trim curves, texture layer
blending, deformers on NURBS surfaces. They are all reported as warnings rather than dropped silently.

This is a big diff. It is one commit here because the pieces lean on each other, but I have it split into ten
self-contained batches on my side (parsing, animation curves, inherit modes, materials, geometry, cameras and
property animation, scene features, NURBS, legacy formats, constraints) and can open them one at a time if that
is easier to review.

The corpus harness, the ufbx oracle, the GPU viewer and the playground replay live here:
https://github.com/Pryme8/babylon-fbx-loader (the corpus files themselves are ufbx's, not committed to Babylon).

Credit where it is due: I built this with Claude Code (Anthropic's Claude) doing a big share of the work with me,
from porting the ufbx semantics to writing the corpus harness and the viewer, while I steered, tested and made the
calls on what goes in. It is listed as a co-author on the commit.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18907/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18907/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18907/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18907/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18907/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18907/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

File: packages/dev/loaders/src/FBX/types/fbxTypes.ts [hunk 1]
```diff
@@ -1,119 +1,150 @@
...
+ * @param value - Parsed FBX object ID value
+ * @returns The object ID, or undefined when the value is not numeric
+ */
+export function getSafeFBXObjectId(value: unknown): number | undefined {
+    if (typeof value !== "number") {
+        return undefined;
+    }
+    // IDs beyond 2^53 lose precision when read as a double, but identical bytes always produce the identical
+    // double, so they remain stable map keys. ufbx accepts such files (e.g. negative 64-bit IDs); so do we.
+    return value;
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/types/fbxTypes.ts:90]: This no longer preserves FBX object identity. Both parsers convert int64 values to JavaScript `number`, so distinct IDs such as `9007199254740992` and `9007199254740993` become the same `Map` key and one object can overwrite the other. The new test exercises only one unsafe ID, so it cannot detect this collision. Please carry IDs losslessly as `bigint`/canonical strings (and reject non-finite or fractional IDs); synthetic negative IDs also need a disjoint namespace.

File: packages/dev/loaders/src/FBX/interpreter/animation.ts [hunk 2]
```diff
@@ -1,852 +1,1213 @@
...
+        return null;
+    }
+
+    // Collect all CurveNodes from all layers
+    const allCurveNodes: FBXCurveNodeData[] = [];
+    const allUnsupportedCurveNodes: FBXUnsupportedCurveNodeData[] = [];
+    const layers: FBXAnimationLayerData[] = [];
+    const diagnostics: FBXAnimationDiagnostic[] = [];
+    let minTime = Infinity;
+    let maxTime = 0;
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/animation.ts]: With no declared time span, a clip whose keys are entirely negative never updates `maxTime` from `0`. For keys at `[-2, -1]`, `finalizeAnimStack()` therefore produces a stop time of `0`, adding an unauthored interval and incorrect duration. The legacy Takes path has the same initializer. Please initialize this to `-Infinity`, handle the no-key case explicitly, and cover both modern and legacy negative-only clips.

File: packages/dev/loaders/src/FBX/interpreter/propertyTemplates.ts [hunk 3]
```diff
@@ -1,169 +1,287 @@
...
+        for (const p of container.children) {
+            if (p.name !== "P" && p.name !== "PS" && p.name !== "Property") {
+                continue;
+            }
+            const name = getPropertyValue<string>(p, 0);
+            if (typeof name !== "string") {
+                continue;
+            }
+            const type = getPropertyValue<string>(p, 1);
+            const flags = getPropertyValue<string>(p, is70 ? 3 : 2);
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/propertyTemplates.ts]: The offsets must follow the property-node syntax, not the container name. A `P` entry inside `Properties60` still has the four metadata fields used by `Properties70`; this reads its flags from the label and includes the real flags in `values`. That shifts axis/unit/material/user properties in transitional 6.x files. Use `p.name === "Property" ? 3 : 4` (and the matching flags index), independently of the container.

File: packages/dev/loaders/src/FBX/interpreter/legacyDocument.ts [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        extra.push(
+            makeNode(
+                "GlobalSettings",
+                [],
+                [makeNode("Version", [int(1000)]), makeNode("Properties60", [], [property60("TimeMode", "enum", "", 14), property60("CustomFrameRate", "double", "", fps)])]
+            )
+        );
+    }
+
+    const kept = doc.nodes.filter((n) => n.name !== "Model" && n.name !== "Takes" && n.name !== "Settings" && n.name !== "GlobalSettings");
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/legacyDocument.ts]: This removes the original `GlobalSettings`, while the replacement above contains only `TimeMode` and `CustomFrameRate`. Legacy `UpAxis`, `FrontAxis`, `CoordAxis`, sign fields, and `UnitScaleFactor` are therefore lost and the converted scene falls back to default orientation/units. Please preserve the existing block and merge only the frame-rate properties.

File: packages/dev/loaders/src/FBX/interpreter/nurbs.ts [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+        name,
+        positions: tess.positions,
+        indices: tess.indices,
+        normals: tess.normals,
+        uvs: tess.uvs,
+        uvSets: [{ name: "map1", data: tess.uvs }],
+        colors: null,
+        tangents: null,
+        binormals: null,
+        controlPointIndices: null,
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/nurbs.ts:675]: A tessellated NURBS surface can still have skin or blend-shape deformers connected by geometry ID, but `controlPointIndices: null` means skin weight generation leaves every vertex at zero weight and blend-shape generation skips the mesh. Skinned NURBS can collapse and NURBS morphs disappear. The tessellator needs to retain/interpolate control-point contributions, or these deformers must be rejected with a diagnostic until supported.

File: packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        if (total > 0) {
+            position.scaleInPlace(1 / total);
+            scale.scaleInPlace(1 / total);
+        }
+        return { position, rotation, scale, total };
+    }
+
+    private _applyWeighted(current: Vector3, desired: Vector3, mask: readonly boolean[]): Vector3 {
+        const w = this.constraint.weight;
+        return new Vector3(
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts]: This blends from the node's previously constrained value on every render. A static node at `0`, target `10`, weight `0.5` becomes `5` on attach, then `7.5`, `8.75`, and eventually converges to the full target. Constraint weight should remain 50%, not become frame-rate-dependent damping. Blend against the freshly evaluated unconstrained transform for the current frame.

File: packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    ) {
+        this.name = `fbxConstraint:${constraint.name}`;
+    }
+
+    public init(): void {}
+
+    public attach(target: TransformNode): void {
+        this.attachedNode = target;
+        const scene = target.getScene();
+        this._observer = scene.onBeforeRenderObservable.add(() => this.evaluate());
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts]: One observer per constraint makes chained constraints depend on FBX object/attachment order: a constraint that reads a target solved by a later observer sees the previous frame. This path also allocates many matrices/vectors/quaternions per constraint per frame despite the scratch object. Please use one scene-level, dependency-ordered solver with explicit cycle handling and ToRef-style reusable temporaries; `constraints: "metadata"` should remain the default until then.

File: packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts [hunk 8]
```diff
@@ -1,3198 +1,4577 @@
...
+
+    private _createAnimationGroup(
+        animStack: FBXAnimationStackData,
+        rigs: FBXRigData[],
+        skeletonByRigId: Map<string, Skeleton>,
+        scene: Scene,
+        modelIdToNode: Map<number, TransformNode>,
+        modelIdToData: Map<number, FBXModelData>,
+        meshes: Mesh[],
+        propertyTargets: { cameraByAttributeId: Map<number, FreeCamera>; lightByAttributeId: Map<number, Light>; materialCache: Map<numbe
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts:2936]: Blend-shape curves are emitted independently per curve node instead of being evaluated through the layer stack. The property-animation loop below has the same issue, and `FBXUnsupportedCurveNodeData` does not retain a layer index at all. Multiple layers animating a morph influence, visibility, camera, light, material, or user property therefore ignore layer modes and weights. Please group by target/property and bake the same layered evaluation used for TRS.

File: packages/dev/loaders/src/FBX/interpreter/materialModel.ts [hunk 9]
```diff
@@ -0,0 +1,856 @@
...
+        if (a === 0x38420192 && b === 0x45fe4e1b) {
+            return { shaderType: "gltfMaterial", shaderPropPrefix: "3dsMax|" };
+        }
+        if (a === 0xd00f1e00 && b === 0xbe77e500) {
+            return { shaderType: "3dsMaxPbrMetalRough", shaderPropPrefix: "3dsMax|main|" };
+        }
+        if (a === 0xd00f1e00 && b === 0x01dbad33) {
+            return { shaderType: "3dsMaxPbrSpecGloss", shaderPropPrefix: "3dsMax|main|" };
+        }
+        if (source.blenderPbr) {
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/materialModel.ts]: This Blender PBR branch is unreachable: `extractMaterial()` never sets `source.blenderPbr`, and even if it did, Blender's documented `Phong` model has already selected `fbxPhong`, while this check is inside `shaderType === "unknown"`. The advertised Blender PBR path therefore remains `StandardMaterial`. Derive the exporter flag from document metadata and apply it before Lambert/Phong classification, with a fixture asserting the resulting PBR values.

File: packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts [hunk 10]
```diff
@@ -1,3198 +1,4577 @@
...
+
+        return container;
+    }
+
+    // ── Parsing ────────────────────────────────────────────────────────────
+
+    /** Parses and interprets the file, reporting the two milestones through the progress callback. */
+    private _parseAndInterpret(data: unknown, onProgress?: (event: ISceneLoaderProgressEvent) => void): FBXSceneData {
+        const doc = this._parse(data);
+        onProgress?.({ lengthComputable: true, loaded: 1, total: 3 });
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts]: `ISceneLoaderProgressEvent.loaded/total` represent byte progress and share the callback used by SceneLoader downloads. Emitting synthetic phase counts after download completion can regress progress from completed bytes to `1/3`, and this event is emitted only after synchronous parsing has already finished. Please preserve byte progress or introduce a separate processing-stage callback rather than mixing units.

File: packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts [hunk 11]
```diff
@@ -1,3198 +1,4577 @@
...
+                }
+            }
+            return meshes;
+        };
+        const base = meshesOf(model.children[0]);
+        for (let level = 1; level < model.children.length; level++) {
+            const levelMeshes = meshesOf(model.children[level]);
+            const threshold = lod.thresholds[level - 1] ?? lod.thresholds[lod.thresholds.length - 1] ?? 0;
+            for (let i = 0; i < base.length; i++) {
+                const target = levelMeshes[i] ?? null;
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts:1020]: `DisplayLevels` are parsed with semantics `0=LOD, 1=show, 2=hide`, but every child is added to the LOD chain here without consulting that mode. Explicitly hidden children can render and always-show children become distance-switched. Apply each child's display mode before registering LOD levels.

File: packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts [hunk 12]
```diff
@@ -1,3198 +1,4577 @@
...
+                material.reflectivityTexture = spec;
+            }
+            const gloss = textureFor(pbr.glossiness) ?? textureFor(pbr.roughness);
+            if (gloss) {
+                material.microSurfaceTexture = gloss;
+            }
+        } else {
+            // Babylon reads metalness and roughness from one texture (B = metal, G = rough). Grayscale maps have
+            // identical channels, so either map works alone; two different maps cannot both be honoured.
+            const metalTex = textureFor(pbr.metalness);
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxFileLoader.pure.ts]: Both texture objects are created before deciding which one Babylon can use. When the maps differ, one wrapper/load is abandoned but remains registered with the scene; in `loadAssetContainerAsync` it may not even be transferred because it is not active on a material. Compare the source references first and instantiate only the selected texture, or explicitly dispose rejected textures.

File: packages/dev/loaders/src/FBX/interpreter/animation.ts [hunk 13]
```diff
@@ -1,852 +1,1213 @@
...
+                            minTime = time;
+                        }
+                        if (time > maxTime) {
+                            maxTime = time;
+                        }
+                    });
+                    layerUnsupportedCurveNodes.push(unsupported);
+                    allUnsupportedCurveNodes.push(unsupported);
+                    const diagnostic: FBXAnimationDiagnostic = {
+                        type: "unsupported-curve-node",
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/animation.ts:223]: This diagnostic is also created for property curves that the loader later imports (visibility, camera, light, material, and user properties), so successful animation is reported as “not evaluated at runtime.” Because diagnostics are dispatched before animation groups are built, `onWarning` cannot distinguish supported mappings. Create this warning only after determining that no runtime property mapping exists.

File: packages/dev/loaders/src/FBX/interpreter/materials.ts [hunk 14]
```diff
@@ -1,213 +1,317 @@
...
+    }
+
+    // LayeredTexture: several textures blended into one slot. Babylon has no layer blending, so the first layer
+    // stands in for the stack; the layer count is recorded on the ref for diagnostics.
+    for (const { id: layeredId, propertyName } of getChildren(objectMap, materialId, "LayeredTexture")) {
+        const layers = getChildren(objectMap, layeredId, "Texture");
+        if (layers.length === 0) {
+            continue;
+        }
+        const ref = extractTextureRef(layers[0].id, layers[0].node, propertyName, objectMap, template);
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/interpreter/materials.ts:192]: Only the first texture layer is used and the remaining layers/order/blend data are discarded. That is a documented approximation in `sceneDiagnostics.ts`, but the PR description currently presents layered textures as supported. Please either implement layer composition or narrow the stated scope and user-facing documentation.

**Review by Popov72 — CHANGES_REQUESTED** [review]
The expansion of FBX support has clear value, but this PR is not ready to merge in its current form.

The change combines parser/legacy conversion, animation evaluation, material policy, geometry/NURBS, scene hierarchy, diagnostics, and a new per-frame constraint system in a single 23k-line diff. The inline findings include data-corrupting ID collisions, incorrect negative clip ranges, broken legacy property/global-setting conversion, non-layered property/morph animation, accumulating constraint weights, and deformed NURBS without control-point mappings.

The default behavior also changes for existing users (`PBRMaterial`, shared geometry, imported curves, runtime constraints, authored animation ranges, and camera/light parenting). These should preserve the previous defaults unless this is explicitly treated as a breaking release with migration guidance.

Please address the findings and add in-repository regression coverage for each claimed behavior. The external ufbx corpus is valuable supporting evidence, but it does not replace pinned CI coverage for the edge cases and public options introduced here. CI is also currently red for side-effects manifest drift and missing TypeDoc comments.

**Comment by Pryme8** [comment]
Thanks for the thorough pass, every one of those was real. Fixed and pushed:

- Defaults are now the 9.26 behaviour. Everything new is an option, bundled as `preset: "full"` (table in the description). No breaking change.
- Ids beyond 2^53 are carried losslessly (exact digits from both parsers, remapped to collision-free ids), fractional/non-finite ids are rejected, synthetic ids never overlap the file's ids.
- Negative-only clips, `P` entries inside Properties60, and legacy GlobalSettings all fixed as you described, with tests for the modern and legacy paths.
- Constraints: one solver per scene, dependency ordered with cycle reporting, no per-frame allocations, and the blend starts from the frame's animated transform so a 50% weight stays 50%. Metadata is the default.
- Morph, visibility, camera, light, material and user property animation go through the layer stack now.
- NURBS surfaces with skins or blend shapes are imported undeformed and reported. LOD display levels are applied before wiring levels. Only the metal/rough (and opacity) texture actually used gets created. The byte progress callback is left alone. Property curves are only reported as unevaluated when there is no mapping.
- Dropped the unreachable Blender PBR branch and narrowed the layered texture wording to "first layer, reported".
- CI: manifest regenerated, TypeDoc comments added, and 27 new unit tests in `test/unit/FBX` pin each of the above plus every public option.

Still one commit for now. If you would rather review it as a series, I have it split into ten batches (parsing, curves, inherit modes, materials, geometry, cameras and property animation, scene features, NURBS, legacy, constraints) and can open them one at a time against master. Your call.

**Review by Popov72 — COMMENTED** [review]
Re-review of bf45b4e6121c8e54a1b4f6576a7a48c27554ab0b: the ID handling, negative clip ranges, Properties60 parsing, legacy GlobalSettings preservation, NURBS deformer fallback, constraint scheduling/allocation work, layered property and morph animation, compatibility defaults, Blender claim, progress callback, LOD modes, texture creation, diagnostics, and layered-texture documentation have all been addressed.

One constraint-weight edge case remains in the new base-capture logic; see the inline comment. CI for the new head has not run yet, so this review does not assess the claimed check results.

File: packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+        node.computeWorldMatrix(true).multiplyToRef(Scratch.invRoot, out);
+        return out;
+    }
+
+    /**
+     * Takes the node's current transform as the unconstrained value unless it is still the previous solve's output.
+     * @param node - Constrained node
+     */
+    private _captureBase(node: TransformNode): void {
+        if (!node.position.equals(this._lastWritten.position)) {
```
Comment (Popov72) [inline packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts]: This equality heuristic cannot tell “animation did not write this frame” from “animation wrote a value equal to the previous constrained output.” For example, after a 50% constraint writes `5`, if the next animation sample also writes an unconstrained value of `5`, this leaves `_base` at its older value (`0`) and solves back to `5` instead of the expected `7.5`. The current test changes the value to `2`, so it misses this case. Please capture the post-animation transform through an explicit animation-phase mechanism rather than inferring writes from value inequality, and add this equality regression case.

Comment (Pryme8) [reply packages/dev/loaders/src/FBX/fbxConstraintBehavior.ts]: Good catch, the equality check was a guess and you found the case where it guesses wrong. Replaced it with an explicit animation-phase mechanism: the solver restores each node's unconstrained transform on `onBeforeAnimationsObservable` and captures + solves on `onAfterAnimationsObservable`, so the blend always starts from what animation produced this frame regardless of the value. Added your exact case as a regression test (sample equal to the previous solved value, expects 7.5) plus a `scene.render()`-driven test with a real `Animation`. Pushed as 32646ce.

**Comment by Pryme8** [comment]
Thank you for hopping on this review, I wanted to try to get this in before the weekend was over since work will take up all my time. @Popov72

**Review by Popov72 — APPROVED** [review]
Re-review of 32646ce133de16c04a11db0e8597aa0a0422ddfe: the remaining constraint base-capture issue is addressed. The solver now restores the unconstrained transform before animation, captures the post-animation value explicitly, and solves afterward; the exact equality edge case and a Scene.render-driven animation path are covered by regression tests. All findings from my review are addressed. CI for this head is still pending.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/loaders/src/FBX/parsers/fbxAsciiParser.ts [hunk 16]
```diff
@@ -1,362 +1,464 @@
...
+ */
+function reinterpretKeyAttrDataFloat(nodes: FBXNode[]): void {
+    for (const node of nodes) {
+        if (node.name === "KeyAttrDataFloat" && node.properties.length === 1 && node.properties[0].value instanceof Float64Array) {
+            const raw = node.properties[0].value;
+            const bits = new Int32Array(raw.length);
+            for (let i = 0; i < raw.length; i++) {
+                bits[i] = raw[i] | 0;
+            }
+            node.properties[0] = { type: "float32[]", value: new Float32Array(bits.buffer) };
```
[the remark this answers is not in the document]
Comment (Popov72) [reply packages/dev/loaders/src/FBX/parsers/fbxAsciiParser.ts:41]: System.Collections.Hashtable.body

File: packages/dev/loaders/src/FBX/interpreter/animation.ts [hunk 17]
```diff
@@ -1,852 +1,1213 @@
...
+    let blended = false;
+    let additive = false;
+    switch (blendMode) {
+        case 0:
+            blended = true;
+            additive = true;
+            break;
+        case 2:
+            blended = true;
+            break;
```
[the remark this answers is not in the document]
Comment (Popov72) [reply packages/dev/loaders/src/FBX/interpreter/animation.ts:645]: System.Collections.Hashtable.body

File: packages/dev/loaders/src/FBX/parsers/fbxAsciiParser.ts [hunk 18]
```diff
@@ -1,362 +1,464 @@
...
+            }
+            this.pos++;
+        }
+        if (this.pos < this.len) {
+            this.pos++; // skip closing quote
+        }
+        // The SDK writes double quotes inside strings as &quot;
+        if (value.includes("&quot;")) {
+            value = value.split("&quot;").join('"');
+        }
```
[the remark this answers is not in the document]
Comment (Popov72) [reply packages/dev/loaders/src/FBX/parsers/fbxAsciiParser.ts:206]: System.Collections.Hashtable.body

**Comment by Pryme8** [comment]
@Popov72 should I do a forum feature update announcement with what was added once this is merged?

**Comment by Popov72** [comment]
> @Popov72 should I do a forum feature update announcement with what was added once this is merged?

I will let @deltakosh answer this one and merge the PR (I think you can't create announcements anymore, only @deltakosh can).

And thanks for your hard work!

**Comment by Pryme8** [comment]
Ok I have one written up for him if he wants to save some time lol.

Also be on the look out for root animation support next.  Its due time.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/18907/merge

Test environment:
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/index.html#WGZLGJ#4600

Links to test your changes to core in the published versions of the Babylon tools (does not contain changes you made to the tools themselves):

https://playground.babylonjs.com/?snapshot=refs/pull/18907/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/18907/merge
https://gui.babylonjs.com/?snapshot=refs/pull/18907/merge
https://nme.babylonjs.com/?snapshot=refs/pull/18907/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/18907/merge#BCU1XR#0

*If you made changes to the sandbox or playground in this PR, additional comments will be generated soon containing links to the dev versions of those tools.*

**Comment by bjsplat** [comment]
Interaction tests

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/testResults/interactionplaywright/index.html

**Comment by bjsplat** [comment]
## 🟢 Memory Leak Test Results

**4** passed, **0** leaked out of **4** scenarios

🟢 All memory leak tests passed — no leaks detected.

<details>
<summary>Passed Scenarios (4)</summary>

| Scenario | Package |
|----------|---------|
| Core Playground #2FDQT5#1508 | @babylonjs/core |
| Core Playground #T90MQ4#14 | @babylonjs/core |
| Core Playground #8EDB5N#2 | @babylonjs/core |
| Core Playground #LL5BIQ#636 | @babylonjs/core |

</details>

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/18907/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
## ⚡ Performance Test Results

🟢 All performance tests passed — no regressions detected.
